bug(completion_contract): Haiku 4.5 tool_choice=any violates Require_tool_use when extended thinking consumes output budget

## Symptom

MASC keeper running on cascade \`vendor_mix_balanced\` fails all tiers with:

\`\`\`
=2331 max_turns=2332 turn_duration_sec=29.814 elapsed_run_sec=29.814
model=claude-haiku-4-5-20251001
stop=error:Completion contract [require_tool_use] violated:
  required tool contract unsatisfied: tool_choice requested tool use,
  but the model returned no ToolUse block
cascade vendor_mix_balanced exhausted: all tiers failed
\`\`\`

Observed 2026-04-18 ~18:04 on live keeper turn 2331/2332.

## Context

- `oas/lib/completion_contract.ml:40-50` — `Require_tool_use` contract errors when response has no `ToolUse` block, regardless of `stop_reason`.
- `oas/lib/llm_provider/capabilities.ml:268-271` — `claude-haiku-4*` inherits `anthropic_capabilities` with `supports_tool_choice=true`, `supports_extended_thinking=true`, `max_output_tokens=8192`.
- `oas/lib/llm_provider/capabilities.ml:158-190` — GLM had **identical** symptom pattern (8+ CompletionContractViolation in single MASC `vendor_mix_balanced` session), resolved by setting `supports_tool_choice=false` so contract relaxes to `Allow_text_or_tool`.

## Hypothesis

"All tiers failed" with the same error signals a **request-side pattern**, not per-model misbehavior. Most likely:

1. **Extended thinking consumes the 8192 output budget before a ToolUse block emits.** `stop_reason` becomes `MaxTokens` or `PauseTurn`. The current validator only checks `content` for `ToolUse` blocks — it does not consider `stop_reason`, so a paused turn with `Thinking` blocks and no ToolUse fails the contract even though the correct action is to continue the turn.

2. **Long-running keeper (turn 2331) + context compression** may have altered the tools payload such that the model cannot select a tool, then defaults to text — observed empirically with GLM and now with Anthropic.

## Proposed fix

**Tier 1 — stop_reason awareness** (preferred):

\`\`\`ocaml
let validate_response ~(contract : t) (response : api_response) :
    (unit, string) result =
  match contract with
  | Require_tool_use
    when response.stop_reason = MaxTokens
      || response.stop_reason = PauseTurn ->
    (* Turn is resumable; bubble up continue signal, not violation. *)
    Ok ()
  | Require_tool_use ->
    if tool_use_names response <> [] then Ok ()
    else Error "..."
  | ...
\`\`\`

This matches the pattern GLM's comment describes: keep `supports_tools=true` but do not error when the model emits text-only under a tool-required request. The stop_reason discrimination narrows it to the specific resumable case, not blanket relaxation.

**Tier 2 — thinking budget clamp**:

When `supports_extended_thinking=true` and the model's `max_output_tokens <= 8192` (i.e. Haiku 4.5), clamp thinking budget to ≤ 50% of max_output so ToolUse emission has headroom. Requires Anthropic backend support for `thinking.budget_tokens` field.

**Tier 3 — per-request override**:

Expose `supports_tool_choice_override` through the consumer API so MASC keeper configs can disable the contract for specific cascades, matching how GLM was handled but at config granularity rather than hardcoded.

## Test plan

- [ ] Reproduce with a synthetic agent that sets `tool_choice=Any`, `extended_thinking=true`, `max_output_tokens=8192`, and a tool whose description is large enough to force thinking.
- [ ] Assert that Tier 1 patch returns `Ok ()` when `stop_reason=MaxTokens`+no ToolUse.
- [ ] Regression: existing \`completion_contract\` tests (17 inline \`let%test\`) must still pass.

## Out of scope

- MASC-side fix for whatever tool_choice injection happens in \`vendor_mix_balanced\` — tracked separately if needed. OAS should behave correctly regardless.
- Changing `supports_tool_choice=true` for anthropic_capabilities wholesale — too broad, GLM's fix was only appropriate because GLM empirically does not honor the field. Anthropic does honor it.

## Log reference

\`\`\`
[2026-04-18 18:04:29] [ERROR] [Misc] cascade vendor_mix_balanced exhausted:
  all tiers failed (last model=claude-haiku-4-5-20251001,
  error=Completion contract [require_tool_use] violated: ...)
\`\`\`

--- a/lib/completion_contract.ml
+++ b/lib/completion_contract.ml
@@ -37,6 +37,20 @@ let tool_use_names (response : api_response) =
       | _ -> None)
     response.content
 
+(* A response whose stop_reason signals the model was cut off mid-turn rather
+   than cleanly deciding not to call a tool. Surfacing a
+   [CompletionContractViolation] in that case is misleading — the caller can
+   continue the turn (or raise max_tokens) instead. Observed on Anthropic
+   Haiku 4.5 where extended thinking consumes the 8192 output budget before
+   a ToolUse block emits: Anthropic returns [pause_turn] which currently
+   parses to [Unknown "pause_turn"] since it is not yet a first-class
+   [stop_reason] variant. *)
+let stop_reason_is_resumable (sr : stop_reason) : bool =
+  match sr with
+  | MaxTokens -> true
+  | Unknown "pause_turn" -> true
+  | EndTurn | StopToolUse | StopSequence | Unknown _ -> false
+
 let validate_response ~(contract : t) (response : api_response) :
     (unit, string) result =
   match contract with
@@ -44,12 +58,15 @@ let validate_response ~(contract : t) (response : api_response) :
   | Require_tool_use ->
     if tool_use_names response <> []
     then Ok ()
+    else if stop_reason_is_resumable response.stop_reason
+    then Ok ()
     else
       Error
         "required tool contract unsatisfied: tool_choice requested tool use, \
          but the model returned no ToolUse block"
   | Require_specific_tool name ->
     (match tool_use_names response with
+     | [] when stop_reason_is_resumable response.stop_reason -> Ok ()
      | [] ->
        Error
          (Printf.sprintf
@@ -246,3 +263,101 @@ let%test "text-only response passes when contract relaxed for unsupported provid
       telemetry = None;
     }
   = Ok ()
+
+(* --- stop_reason resumability tests (extended-thinking / budget cut-off) --- *)
+
+let%test "Require_tool_use accepts no-ToolUse response when stop_reason is MaxTokens" =
+  validate_response
+    ~contract:Require_tool_use
+    { id = "r";
+      model = "claude-haiku-4-5-20251001";
+      stop_reason = MaxTokens;
+      content = [ Text "partial thinking output" ];
+      usage = None;
+      telemetry = None;
+    }
+  = Ok ()
+
+let%test "Require_tool_use accepts no-ToolUse response when stop_reason is Unknown pause_turn" =
+  validate_response
+    ~contract:Require_tool_use
+    { id = "r";
+      model = "claude-haiku-4-5-20251001";
+      stop_reason = Unknown "pause_turn";
+      content = [ Text "thinking..." ];
+      usage = None;
+      telemetry = None;
+    }
+  = Ok ()
+
+let%test "Require_tool_use still rejects no-ToolUse response when stop_reason is EndTurn" =
+  match
+    validate_response
+      ~contract:Require_tool_use
+      { id = "r";
+        model = "m";
+        stop_reason = EndTurn;
+        content = [ Text "I decline to call tools" ];
+        usage = None;
+        telemetry = None;
+      }
+  with
+  | Error _ -> true
+  | Ok () -> false
+
+let%test "Require_tool_use still rejects no-ToolUse response when stop_reason is Unknown refusal" =
+  match
+    validate_response
+      ~contract:Require_tool_use
+      { id = "r";
+        model = "m";
+        stop_reason = Unknown "refusal";
+        content = [ Text "..." ];
+        usage = None;
+        telemetry = None;
+      }
+  with
+  | Error _ -> true
+  | Ok () -> false
+
+let%test "Require_specific_tool accepts no-ToolUse response when stop_reason is MaxTokens" =
+  validate_response
+    ~contract:(Require_specific_tool "calculator")
+    { id = "r";
+      model = "m";
+      stop_reason = MaxTokens;
+      content = [ Text "partial" ];
+      usage = None;
+      telemetry = None;
+    }
+  = Ok ()
+
+let%test "Require_specific_tool still rejects wrong-tool response when stop_reason is MaxTokens" =
+  match
+    validate_response
+      ~contract:(Require_specific_tool "calculator")
+      { id = "r";
+        model = "m";
+        stop_reason = MaxTokens;
+        content =
+          [ ToolUse
+              { id = "call-1";
+                name = "search";
+                input = `Assoc [];
+              }
+          ];
+        usage = None;
+        telemetry = None;
+      }
+  with
+  | Error _ -> true
+  | Ok () -> false
+
+let%test "stop_reason_is_resumable classification" =
+  stop_reason_is_resumable MaxTokens
+  && stop_reason_is_resumable (Unknown "pause_turn")
+  && not (stop_reason_is_resumable EndTurn)
+  && not (stop_reason_is_resumable StopToolUse)
+  && not (stop_reason_is_resumable StopSequence)
+  && not (stop_reason_is_resumable (Unknown "refusal"))
+  && not (stop_reason_is_resumable (Unknown "anything-else"))

